main/src/ca/uwaterloo/flix/api/lsp/provider/HoverProvider.scala

```diff
@@ -28,16 +30,46 @@ import scala.annotation.tailrec
 
 object HoverProvider {
 
-  def processHover(uri: String, pos: Position)(implicit index: Index, root: Root, flix: Flix): JObject = {
-    index.query(uri, pos) match {
-      case None => mkNotFound(uri, pos)
+  def processHover(uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = {
+    var stack: List[AnyRef] = Nil
+    def push(x: AnyRef): Unit = {
+      stack = x :: stack
+    }
+    object HoverConsumer extends Visitor.Consumer {
```

Comment:
I think this can be encapsulated in one reusable abstraction. 

A class that both extends Consumer and has the stack inside it.

Reply:
Yeah, we could do that, but wouldn't we then need to implement each consumer such that everything is added to the stack? We might not want that. In this casee in particular, I don't think that's useful. I guess we could add everything to the stack and just ignore the things we don't care about by looking at the next element if the head is "uninteresting" to us.

Reply:
I don't understand. I am just saying that the current code can be encapsulated in a reusable class.

Reply:
Like you will need the same functionality in e.g. HighlightProvider.

Reply:
"StackConsumer"

Reply:
My problem with the abstraction is that what needs to be added to the stack is dependent on the use case. What we want to add to the stack and what we want to ignore, might be different between HoverProvider and HightlightProvider. Then this abstraction would need to still provide the same default implementations as the Consumer, so that each use site may implement just the consumers they need. But at that point, the abstraction gives us nothing, except maybe packaging the stack itself within the consumer behind some interface

Reply:
Unless ofc. we just want everything on the stack. Then yeah, that abstraction would make sense

Reply:
The HighlightProvider will need to visitors, but it seems to me that the first one would very much be exactly equivalent to this one-- i.e. to find the item that the cursor is on.

Reply:
What I'm saying is, we might not want to add every element visited to the stack. I don't think it's useful to add all the elements visited to the stack in the hover provider for instance. But what that means is that such a StackConsumer can't actually implement any consumers, since that would lock-in what's added to the stack. Only if we always wish to add everything visited to the stack, does this abstraction then make sense

Reply:
any if StackConsumer can't implement ant consumeX functions, then StackConsumer adds very little if anything

Reply:
Do you agree that HoverProvider and HighlightProvider need to perform the same search?

Reply:
their searches will visit the same elements, but does that necessarily mean they want the same consumers?

Reply:
I suppose, at the end of the day, it's only the very last (most specific) element that matters in both cases. But even here, there might be reasons that we want the two cases to be different. Imagine that that you're hovering on some element `x` but actually within that element is a subelement `y` that you're more specifically hovering on. Now lets say that the hover can't say anything about type of AST node. Then we might want to "backtrack" and actually give information about the parent `x`. We could add it all to the stack and backtrack, or, if we just never add the elements we can't say anything about, we get the same effect but without having to actively backtrack. Now lets say we're highlighting `x`, maybe in this case highlighting the most specific `y` object does make sense, and then we don't want to backtrack. We can ofc. use the same Consumer that adds everything visited (including `y`) to the stack in both cases, and backtrack in the hover provider case. But instead we could simply use two different Consumers. For the hover provider we could use one that implements consumeX and but doesn't implement consumeY, so that y (type Y) is just never added to the stack. For the hightlight provider, we could then implement both consumeX and consumeY, such that y is added to the stack. This is what I mean by different situations maybe calling for different things to be added to the stack. With a `StackConsumer`, we'd have to add everything to the stack and backtrack if necessary.

Reply:
But maybe its better to just always find the most specific element and just backtrack later?

Reply:
In that case, then yeah, the StackConsumer makes perfect sense

Reply:
sorry for the text wall ^^'

Reply:
Am I just over complicating it again?

Reply:
YAGNI. You are designing for what-ifs

Reply:
In both cases you may consider something simpler than a stack: Just keep one element.

Reply:
But I think it would be best with a stack that can consume everything. That seems very reusable.

Reply:
Yeah, I think you're right. That is probably better

Reply:
Should I add this in a new PR since its a change in HoverProvider.scala, or is it fine to do it as a part of this PR?